lib/uklibid/Config.uk

```diff
@@ -1,10 +1,113 @@
-config LIBUKLIBID
+menuconfig LIBUKLIBID
 	bool "uklibid: Library identifier"
-	default n
+	default y
 	# FIXME: We actually depend on <uk/assert.h> but lib/ukdebug also
 	#        depends on us. If we specify the dependency here, it creates a
 	#        circular dependency that KConfig doesn't like. However, we
 	#        can currently assume that lib/ukdebug is always part of a
 	#        build, so we comment out the dependency here. Of course, this
 	#        is not accurate.
 	#select LIBUKDEBUG
+
+if LIBUKLIBID
+config LIBUKLIBID_INFO
+	bool "Embedded library metadata"
+	default y
+
+	if LIBUKLIBID_INFO
+	menu "Global metadata"
+		config LIBUKLIBID_INFO_UKFULLVERSION
+			bool "Full unikraft version"
+			default y
+			help
+				If enabled, the full Unikraft version including
+				Git SHA is embedded. Otherwise, only the short
+				Unikraft version is stored.
+
+		config LIBUKLIBID_INFO_UKCODENAME
+			bool "Unikraft codename"
+			help
+				If enabled, the Unikraft release codename is
+				stored as part of the global metadata
+				information.
+
+		config LIBUKLIBID_INFO_UKCONFIGGZ
+			bool "Embed gzipped configuration (.config)"
+			help
+				If enabled, the current configuration `.config`
+				of the build is gzipped and embedded to the
+				global metadata infomration space.
+	endmenu
+
+	menu "Per library metadata"
+		config LIBUKLIBID_INFO_LIB_UKVERSION
+			bool "Unikraft version"
+			help
+				If enabled, the Unikraft version is included
+				to each library.
+
+		config LIBUKLIBID_INFO_LIB_UKFULLVERSION
+			bool "Full version"
+			depends on LIBUKLIBID_INFO_LIB_UKVERSION
+			help
+				Instead of the short Unikraft version, the full
+				Unikraft version including Git SHA is embedded
+				to each library.
+
+		config LIBUKLIBID_INFO_LIB_UKCODENAME
+			bool "Unikraft codename"
+			help
+				If enabled, the Unikraft release codename is
+				stored as part of each library metadata
+				information.
+	endmenu
+
+	config LIBUKLIBID_INFO_COMPILER
+		bool "Compiler (CC) information"
+		default y
+		help
+			Enables storing the C compiler ($(CC)) that was used.
+
+	config LIBUKLIBID_INFO_COMPILEDATE
+		bool "Compile date"
+		help
+			This option stores the compile date and time for each
+			library and the unikernel image.
+
+	config LIBUKLIBID_INFO_COMPILEDBY
+		bool "Compiled by"
+		help
+			Name of the person/entity that did the compilation.
```

Comment:
This is now automatically set to `user@host`, so maybe the help string can be updated accordingly.

Also when looking at the generated menu via `menuconfig` it looks more like the `Compiled by` is a field that enables the ones right below that set it to a custom value, rather than a field that automatically generates something. That is because the `LIBUKLIBID_INFO_COMPILEDBY` option is overloaded with two semantics: 1. It enables `LIBUKLIBID_INFO_COMPILEDBY_TEXT`, which `depends` on it, and 2. Sets the default value. I think these cases are handled with a `menu` construct, ie:

```
menu "Compiled by"

config LIBUKLIBID_INFO_COMPILEDBY
    bool "user@hostname"
    help

config LIBUKLIBUD_INFO_COMPILEDBY_TEXT
    string "custom"
    help

endmenu
```

PS: Notice in the above example I only used a single option for the custom field, as one field is enough for the user to provide the full string of `user@host` or even just an alternative format they may prefer to use (eg a single organization name).

Reply:
👍🏻  I agree, I am going to change it. Please check the update: I will add the "Compiled by" option as a `menuconfig` submenu.